feat: add lockFactory config

sublimator opened the pull request:
We are using SuperTokens with an extension that needs to refresh the tokens for a given domain that may be very infrequently browsed, or on the other hand may be browsed concurrently with the extension usage. 

We want to use supertokens-website from the background page of the extension, overriding the CookieHandler (to set tokens on the app domain using Web Extension apis) and customizing browser-tabs-tokens so that the extension can act as "just another tab" 

Wanted to get a feel if this PR is moving in the right direction and would have a chance to be merged. 

Thanks!

rishabhpoddar commented:
Thanks for this PR @sublimator .

@porcellus will review this in the coming days.

porcellus commented on lib/ts/utils/windowHandler/index.ts:
```diff
@@ -12,38 +12,45 @@
+    // Declare subset of WindowHandlerReference that is actually used,
+    // so implementors don't need to over-engineer.
+    windowHandler: {
+        location: Pick<
```
We shouldn't limit this to a subset. This lib exports the window handler, and the other props are used by supertokens-web-js and supertokens-auth-react.

sublimator replied:
Fair, I shouldn't have committed this

porcellus commented on lib/ts/utils/windowHandler/index.ts:
```diff
@@ -12,38 +12,45 @@
 
     constructor(windowHandlerInput?: WindowHandlerInput) {
-        let windowHandlerFunc: WindowHandlerInput = original => original;
+        let windowHandlerFunc: WindowHandlerInput = original => original
```
These updates do not match the coding style used elsewhere in the project; please run `npm run build-pretty`.
Also, I'm not sure if you've installed the commit hooks (check [here](https://github.com/supertokens/supertokens-website/blob/master/CONTRIBUTING.md#development-setup) if not)

sublimator replied:
<img width="319" alt="image" src="https://user-images.githubusercontent.com/525211/212615170-6f15c569-a5db-44e8-912a-2976dec81cd9.png">

I think prettier might need to be updated ? Seems this is using 1.x while 2.x is out.

porcellus commented on lib/ts/types.ts:
```diff
@@ -39,6 +40,13 @@ export type InputType = {
     apiDomain: string;
     apiBasePath?: string;
+    /**
+     * Rather than using browser-tabs-lock, allow for a Lock factory to be configured
+     * via SuperTokens.init(...). This can be used for example by a WebExtension that
+     * needs to update cookies for a domain that may or may not have an associated
+     * tab open.
```
```suggestion
     * This allows for a Lock factory to be configured, which defaults to browser-tabs-lock.
     * This can be used, for example, by a WebExtension that needs to update cookies for
     * a domain that may or may not have an associated tab open.
```

porcellus reviewed (changes requested):
This is a good change; there are just a few minor issues :)

sublimator commented:
@porcellus 

I removed the limiting of the windowHandler type, ran prettier, updated it when it wasn't working with some of the syntax I used (`??` operator and `import type`), then tweaked the .prettierrc to be as close to 1.x as possible.   

If you are the maintainer (or one of), feel free to take over PR if it's more convenient for you.  

We just need some ability to customize the lock.

Thanks!

sublimator commented:
I made the lockFactory async. I'm not sure if we will need it, but it's more flexible, so why not?

porcellus reviewed (approved):
LGTM :)

rishabhpoddar commented on package.json:
```diff
@@ -20,7 +20,7 @@
     "mocha": "6.1.4",
```
- Please update the changelog under the [unreleased] section
- Might want to expose the LockFactory in the `/utils/lockFactory` folder similar to how the [windowHandler is exposed](https://github.com/sublimator/supertokens-website/blob/nd-add-lock-factory-to-config-2023-01-13/utils/windowHandler/index.js)

rishabhpoddar replied:
The other changes look good. So once these two comments are addressed, i'll merge and release this.

sublimator commented:
@rishabhpoddar 

I'm not sure how/why I set the target branch to 15.0 and not master ?
Is that right? 

Anyway, I (hopefully) addressed your two comments.
